### DISCUSSION TITLE
Half Duplex on the 1.4.10 tag

### DISCUSSION THREAD

**Discussion by coburnw** [SpenceKonde/megaTinyCore#847] [Show and tell]
I have been toying with the 1.4.10 tag of DxCore with some pretty good success.  Thank you!

In the process of bringing up the RS485 link i found that the half duplex portion of the C serial ISR didn't make it into the assembly language ISR.  These are the changes i found were needed to re-enable the full featured ISR and configure the serial port and _state variable for open drain mode.  These changes are virtually untested so both standard and non-standard disclaimers apply...

* checkout the 1.4.10 tag
* create a file ``platform.local.txt`` in the same folder as platform.txt
  * add ``compiler.cpp.extra_flags=-DUSE_ASM_TXC=0 -DUSE_ASM_RXC=0`` to platform.local.txt
* add the following stanza to UART.cpp just before [`` ctrla &= 0x2B;``](https://github.com/SpenceKonde/DxCore/blob/b0bb30f8cb703991d3fe97a3241091a221eab09b/megaavr/cores/dxcore/UART.cpp#L384) or line 384
```
   #if USE_ASM_TXC != 1                       // OpenDrain mode only supported by C ISR
     if (ctrla & 0x04){                       // see if user wants OpenDrain
       ctrlb |= USART_ODME_bm;
     }
   #endif
```

**Comment by SpenceKonde** [comment]
This is a showstopper bug effecting mTC and DxC equally

The release is otherwise ready except for this critical bug. (it's critical because **I** need it on a 4k dual uart chip)
A non-asm solution is not acceptable. 

I amm, going to try to move this to megatinycore where the issue will be fixed first, and turn it into an issue. This is a show stopper becuause with this broken, my next project is dead.

**Comment by SpenceKonde** [comment]
I also think you are misdiagnosing this. 

ODME mode is not changed in the assembly, it is set in begin and left that way. 

What I did notice, though is that the sender is getting the second and last character that they sent. That is not intended and is unacceptable. This must be fixed.

**Reply by coburnw** [reply]
Ha.  Now **thats** no surprise!

To be clear, half duplex rs485 with the C ISR enabled is working perfectly well for me.  Thank you.

**Comment by SpenceKonde** [comment]
I think I see what is happening here. 

We need to keep track of how many characters are sent that are irrelevant and should be discarded. this is a major rearchitecture of that mechanism that will require a lot of work to fix. It will not halt the release cycle.

**Comment by SpenceKonde** [comment]
After an incredible amount of pain, this was fixed on megaTinycore.
